### ISSUE TITLE
[Ascend][P1] flag_gems cat/concat 多张量路径触发 Triton-Ascend 编译器段错误（VL 并发 chunked-prefill 场景必崩）

### ISSUE THREAD

**Issue by JosephNew** [flagos-ai/FlagGems#6500]
## 现象

昇腾 910C，`flag_gems` rc2 分支 HEAD `4af1e434`（flagtree 0.6.0 + triton 3.5 线），
Qwen3.6-27B **多模态并发**（`qwen3_6_27b_concurrent.py --mode vl`，TP=2，16 并发）时
scheduler 进程 **SIGSEGV 段错误**（exit -11）：

```
Fatal Python error: Segmentation fault
Current thread (most recent call first):
  File ".../triton/backends/ascend/compiler.py", line 124 in ttir_to_linalg   ← 编译器本体段错误
  File ".../triton/backends/ascend/compiler.py", line 830 in <lambda>
  File ".../triton/backends/ascend/spec/triton/compiler/compiler.py", line 285 in compile
  File ".../triton/experimental/tle/__init__.py", line 35 in tle_compile
  File ".../triton/backends/ascend/spec/triton/runtime/jit.py", line 623 in run
  File "/root/FlagGems/src/flag_gems/ops/cat.py", line 181 in _cat_run_kernel   ← flag_gems cat kernel
  File "/root/FlagGems/src/flag_gems/ops/cat.py", line 344 in cat
  File "/root/FlagGems/src/flag_gems/ops/concat.py", line 41 in concat
  File ".../sglang/srt/managers/mm_utils.py", line 653 in _get_chunked_prefill_embedding
  File ".../sglang/srt/managers/mm_utils.py", line 728 in get_embedding_and_mask
  File ".../sglang/srt/managers/mm_utils.py", line 822 in embed_mm_inputs
  File ".../sglang/srt/managers/mm_utils.py", line 1052 in general_mm_embed_routine
  File ".../sglang/srt/models/qwen3_vl.py", line 1272 in forward
  ...
[2026-09-20 15:52:54] Subprocess scheduler_0 (pid=3954860) crashed with exit code -11. Triggering SIGQUIT for cleanup...
```

## 关键结论：与 shape 无关，**多张量 cat 在本环境 100% 编译段错误**

我用真机日志抓到的真实输入 shape，在脱离 sglang 的最小脚本里扫了一遍
（`torch.concat(parts, dim=0)`，bf16，hidden=5120，flag_gems.enable()）：

| 用例 | shapes（各张量行数） | 结果 |
| --- | --- | --- |
| 真机 VL 并发原始调用（15 张量） | 300,100,64,64 ×3 + 300,100,64 | ❌ **SIGSEGV** |
| 单批 4 张量 | 300,100,64,64 | ❌ **SIGSEGV** |
| 单批 3 张量 | 300,100,64 | ❌ **SIGSEGV** |
| 单批 2 张量 | 300,100 | ❌ **SIGSEGV** |
| **单张量** | 300 | ✅ PASS（走 single 分支，不编译 kernel） |
| 两批各 4 张量 | 300,100,64,64 ×2 | ❌ **SIGSEGV** |
| 4 张量**同 shape** | 256,256,256,256 | ❌ **SIGSEGV** |
| 4 张量大方阵 | 1024,1024,1024,1024 | ❌ **SIGSEGV** |

→ **只要走到 `cat_copy_func_kernel`（多张量路径）就崩，跟 shape、跟张量个数的奇偶都无关**；
单张量因为 `_cat_build_working_list` 返回 `mode="single"` 直接返回、不编译 kernel，所以侥幸不崩。

也就是说：**flag_gems 的 `cat`/`concat` 多张量路径在昇腾本环境从未真正可用**，
只是之前没有覆盖到这条路径 —— 单请求 VL 冒烟走的是另一条分支（不触发 chunked-prefill 的 embedding 拼接）。

## 环境

- FlagGems：rc2 分支 HEAD `4af1e434`（含 #6455 等全部已合入修复）
- flagtree：`0.6.0+ascend3.2`（triton 3.5 线）
- sglang：0.5.11 / sgl_kernel_npu 2026.5.1 / plugin 0.2.0-rc2.post1
- 镜像：tree_0.6.0_ascend3.2 老版本镜像（CANN 8.5.0）
- 黑名单：`ascend.yaml` 22 项（含 7 项广播算子，绕 #6238）
- 硬件：昇腾 910C，2 卡 TP=2

### 1) 全链路（约 2 分钟，最快）

```bash
# 昇腾 910C，flagtree 0.6.0 + triton 3.5 + flag_gems rc2 HEAD 4af1e434
export ASCEND_RT_VISIBLE_DEVICES=12,13
export MODEL_PATH=/models/Qwen3.6-27B
export TP_SIZE=2
python3 -u examples/qwen3_6_27b_concurrent.py --mode vl     # → SIGSEGV / RC=137
```

⚠ 只有 **`--mode vl`（并发）** 会崩；单请求 VL 冒烟（`qwen3_6_27b_offline_inference.py`）不触发。
⚠ 只有 **多模态** 会崩；纯文本并发（`--mode text`）不走这条路径，不受影响。

### 2) 脱离 sglang 的最小复现（单次约 18s，不需要加载模型）

> 已自验：逐字跑本段 **3/3 稳定 SIGSEGV**；注释掉 `flag_gems.enable()` 后 **3/3 PASS**（对照）。
> 单次约 17~18s，主要花在 `import torch_npu` + device 初始化上，崩溃发生在编译阶段。

```python
import torch
import torch_npu
import flag_gems

flag_gems.enable()          # 关键：import flag_gems 不注册 op，必须显式 enable
torch_npu.npu.set_device(0) # ⚠ 0 是「可见设备」序号，运行前设 ASCEND_RT_VISIBLE_DEVICES=<空闲卡>

parts = [torch.randn(n, 5120, dtype=torch.bfloat16, device="npu")
         for n in (256, 256, 256, 256)]     # 同 shape 也崩
torch.concat(parts, dim=0)  # → Segmentation fault (exit -11)
```

---

## 影响面

凡是走 **多模态 + chunked prefill** 的 sglang 模型（`mm_utils._get_chunked_prefill_embedding`
结尾的 `torch.concat(embedding_list, dim=0)`），在本环境必然 SIGSEGV。
与 #6249 / PR #6455（linear）**无关**，是独立问题。

## 绕法（已真机验证）

在昇腾 `sglang_fl/dispatch/config/ascend.yaml` 的 `flagos_blacklist` 里加：

```yaml
  - cat      # torch.cat    → cat.py
  - concat   # torch.concat → concat.py:41
```

两个注册名都要加（`flag_gems/__init__.py` 里 `("cat", cat)` 和 `("concat", concat)` 分别注册，
sglang `mm_utils` 走的是 `torch.concat`）。

加上之后，`--mode vl` 从 **SIGSEGV** 变为 **`vl validation passed.`**，
四图全对（red_square→'Red' / cat→'cat' / stop_sign→'red' / digit_seven→'7'），`All concurrent tests passed.`。

## 备注

- 段错误发生在 **Triton-Ascend 编译器 `ttir_to_linalg` 降级阶段**，`.ttir` 已成功生成、
  但 `.json`/`.npubin` 没有产物 —— 可在 `$TRITON_CACHE_DIR` 下看到只剩一个 `cat_copy_func_kernel_4.ttir`。
  同一个 kernel（`BLOCK_X=1024`）在 `--mode text` 全程没被触发过。
- 怀疑是 TTIR → Linalg 降级对 `cat_copy_func_kernel_4` 里那套
  `pid_y == 0/1/2/else` 嵌套 `scf.if` + `idx // (dim_size_in * dim_prod_post)` 结构处理有问题
  （未进一步缩小到具体 op，需要编译器侧同学看）。

**Comment by JosephNew** [comment]
## 补充：最小复现已自验（逐字原文跑，3/3 稳定复现 + 对照组正常）

按「交付的复现命令必须能自己跑出来」的原则，把 issue 正文 §2 那段代码**逐字**（不做任何等价变形、
不钉 `TRITON_CACHE_DIR`/`FLAGGEMS_CACHE_DIR`，用默认缓存）跑了一遍，每组 3 次：

| 组 | 内容 | 结果 |
| --- | --- | --- |
| **A** | **交付原文（逐字）** | **❌ SIGSEGV（exit -11）× 3/3，稳定** |
| **B** | 原文 + 末尾哨兵 `print` | ❌ SIGSEGV（exit -11）× 3/3 —— 确认是**真崩在 `torch.concat`**，不是静默退出 |
| **C** | 对照组：注释掉 `flag_gems.enable()` | **✅ PASS × 3/3** —— 不 enable 就不崩，坐实是 **flag_gems 的 `cat` 路径**问题 |

自验脚本：`/tmp/selfcheck_6500.py`，运行方式
`ASCEND_RT_VISIBLE_DEVICES=12 python3 /tmp/selfcheck_6500.py`

### ⚠️ 两处订正

1. **不是「秒级」，单次约 17~18 秒**。原文写的「秒级编译前」不准确 ——
   这 17s 基本花在 `import torch_npu` + device 初始化上，真正崩溃发生在编译阶段。
   仍比跑全链路（~2 分钟）快很多，但请按 ~18s 预期。
2. **需要显式指定卡**：脚本里 `torch_npu.npu.set_device(0)` 的 `0` 是**可见设备序号**，
   所以运行前需要 `ASCEND_RT_VISIBLE_DEVICES=<空闲卡>`；否则 `device 0` 会落到物理 0 号卡上。

### 附：真机日志抓到的原始调用 shape（供参考）

给 `cat.py` 的 `_cat_run_kernel` 插桩跑一轮 `--mode vl` 拿到的真实输入：

```
dim=0 out_shape=[2048, 5120] n=15
shapes=[(300,5120),(100,5120),(64,5120),(64,5120)] × 3 + [(300,5120),(100,5120),(64,5120)]
```

用这组 shape 以及其它各种 shape（2/3/4 张量、同 shape 256×4、大方阵 1024×4）跑最小复现**全部 SIGSEGV**；
只有**单张量** PASS（走 `_cat_build_working_list` 的 `mode="single"` 分支，根本不编译 kernel）。
故结论不变：**flag_gems `cat`/`concat` 的多张量路径在本环境 100% 编译段错误，与 shape 无关。**

**Comment by JosephNew** [comment]
## 团队评审结论：本 issue 本期不修，昇腾线以黑名单兜底（2026-09-21）

今天团队评审后给出结论，同步到这里：

> 这个算子 gems 修复影响面较大，而且不合适。编译器解决依赖昇腾厂商，暂时黑名单解决，版本结束后再推动修复。

落到本 issue：

- **现象与定位已经很清楚**（见 issue 正文）：崩在 `triton/backends/ascend/compiler.py:124 ttir_to_linalg`，即 **Triton-Ascend 编译器本体段错误**，被编译的是 flag_gems `cat.py:181 _cat_run_kernel`。与 shape 无关、与张量个数奇偶无关，只要走多张量路径就 100% 崩；单张量走 `_cat_build_working_list` 的 `mode="single"` 分支、不编译 kernel，所以侥幸不崩。
  - 触发路径是 **chunked-prefill 的多模态 embedding 拼接**（`mm_utils.py:653 _get_chunked_prefill_embedding`），**单请求 VL 冒烟走不到这条路径**，所以此前「VL 冒烟四图全对」没覆盖到。
- **gems 侧不做修复**。`cat`/`concat` 多张量 codegen 的改动同样落在通用层，影响面大，不是本版本窗口该动的落点。
- **根因在编译器侧**。段错误发生在 Triton-Ascend 的 MLIR lowering 阶段（`.ttir` 已生成、无 `.json`/`.npubin` 产物），需要昇腾厂商侧排查定位 —— **本期依赖厂商**。
- **本期做法**：`ascend.yaml` 的 `flagos_blacklist` 加 `cat` + `concat`（**两个注册名都要加**，`flag_gems/__init__.py` 里 `("cat", cat)` 与 `("concat", concat)` 分别注册，sglang 走 `torch.concat`）。
  - 已验证：加上后 `--mode all` 从 SIGSEGV 变为 **RC=0 / `All concurrent tests passed.`**（text 113.93s / vl 8.10s / mixed 172.12s，四图全对）。黑名单 22 → 24 项。
- **本 issue 保持 open，不关闭**。版本结束后再推动修复，届时撤 `cat`/`concat` 黑名单做回归。

issue 正文里的最小复现（`flag_gems.enable()` + 多张量 `torch.concat`，无需加载模型）经逐字自验：3/3 稳定 SIGSEGV；注释掉 `flag_gems.enable()` 作对照 3/3 PASS。单次约 18s（不是「秒级」，主要花在 `import torch_npu` + device 初始化），供后续推动编译器侧时直接使用。
